### PR TITLE
POD-779 | Initial integration of devpod-crane

### PR DESCRIPTION
**Pull request by janekbaraniewski**
Part of POD-779

### CODE REVIEW

File: cmd/up.go [hunk 1]
```diff
@@ -342,17 +343,19 @@ func (cmd *UpCmd) devPodUp(
 	// get result
 	var result *config2.Result
 
-	// check what client we have
-	if workspaceClient, ok := client.(client2.WorkspaceClient); ok {
-		result, err = cmd.devPodUpMachine(ctx, devPodConfig, workspaceClient, log)
+	switch client := client.(type) {
```
Comment (bkneis) [inline cmd/up.go:346]: nice!

File: pkg/devcontainer/crane/run.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	craneSigningKey string
+)
+
+const (
+	PullCommand    = "pull"
+	DecryptCommand = "decrypt"
+
+	GitCrane = "git"
+
+	BinPath = "devpod-crane" // FIXME
```
Comment (bkneis) [inline pkg/devcontainer/crane/run.go:24]: is this comment still valid?

File: pkg/devcontainer/crane/run.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	Files map[string]string `json:"files"`
+}
+
+func IsAvailable() bool {
+	_, err := exec.LookPath(BinPath)
+	return err == nil
+}
+
+func runCommand(command string, args ...string) (string, error) {
+	cmd := exec.Command(BinPath, append([]string{command}, args...)...)
```
Comment (bkneis) [inline pkg/devcontainer/crane/run.go:39]: why not simply `cmd := exec.Command(BinPath, command, args...)`?

File: pkg/devcontainer/crane/run.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+	var outBuf, errBuf bytes.Buffer
+	cmd.Stdout = &outBuf
+	cmd.Stderr = &errBuf
+
+	err := cmd.Run()
```
Comment (bkneis) [inline pkg/devcontainer/crane/run.go:45]: nit: this could be a one liner

File: pkg/devcontainer/crane/run.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	BinPath = "devpod-crane" // FIXME
+
+	tmpDirTemplate = "devpod-crane-*"
+)
+
+type Content struct {
+	Files map[string]string `json:"files"`
+}
+
+func IsAvailable() bool {
```
Comment (bkneis) [inline pkg/devcontainer/crane/run.go:33]: docbock please for public function :pray:

File: pkg/devcontainer/crane/run.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+	err := cmd.Run()
+	if err != nil {
+		return "", fmt.Errorf("failed to execute command: %v, error: %w", errBuf.String(), err)
+	}
+
+	return outBuf.String(), nil
+}
+
+func PullConfigFromSource(configSource string, log log.Logger) (string, error) {
```
Comment (bkneis) [inline pkg/devcontainer/crane/run.go:53]: docbock please for public function :pray:

**Review by bkneis — COMMENTED** [review]
Looking good! I really like some of the refactoring done here and the new code looks good. I just had a couple of comments / questions then ready to merge

File: pkg/devcontainer/crane/run.go (hunk 2 above)
Comment (janekbaraniewski) [reply pkg/devcontainer/crane/run.go:24]: yeah would be nice to make it at least somehow configurable to support dev/debug binaries

File: pkg/devcontainer/crane/run.go (hunk 3 above)
Comment (janekbaraniewski) [reply pkg/devcontainer/crane/run.go:39]: oh yeah now I remember: 
```
pkg/devcontainer/crane/run.go:40:40: too many arguments in call to exec.Command
        have (string, string, []string)
        want (string, ...string)
```

File: pkg/devcontainer/crane/run.go (hunk 4 above)
Comment (janekbaraniewski) [reply pkg/devcontainer/crane/run.go:45]: thx!

File: pkg/devcontainer/run.go [hunk 7]
```diff
@@ -116,67 +110,57 @@ func (r *runner) Up(ctx context.Context, options UpOptions, timeout time.Duratio
...
+		r.Log.Warn("dev container config is missing one of \"image\", \"dockerFile\" or \"dockerComposeFile\" properties, defaulting to auto-detection")
 
-		result, err = r.runSingleContainer(ctx, substitutedConfig, substitutionContext, options, timeout)
+		lang, err := language.DetectLanguage(r.LocalWorkspaceFolder)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("could not detect project language and dev container config is missing one of \"image\", \"dockerFile\" or \"dockerComposeFile\" properties")
 		}
+
+		if language.MapConfig[lang] == nil {
```
[the remark this answers is not in the document]
Comment (janekbaraniewski) [reply pkg/devcontainer/run.go:137]: yeah definitely. didn't really go into this code yet, tried to break it down a bit to make it easier to refactor.

File: cmd/up.go (hunk 1 above)
Comment (janekbaraniewski) [reply cmd/up.go:346]: thx! 🙇

File: pkg/devcontainer/crane/run.go (hunk 6 above)
Comment (janekbaraniewski) [reply pkg/devcontainer/crane/run.go:54]: added 👍

**Review by pascalbreuninger — APPROVED** [review]
LGTM

**Pull request merged**